File: app/src/main/java/com/example/alarm/AlarmFragment.java
package com.example.alarm;

import android.annotation.SuppressLint;
import android.icu.util.Calendar;
import android.icu.util.ULocale;
import android.os.Bundle;
import android.util.Log;
import android.util.Pair;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import androidx.activity.OnBackPressedCallback;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.content.ContextCompat;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.LinearLayoutManager;

import com.example.alarm.databinding.FragmentAlarmBinding;
import com.example.alarm.model.Alarm;
import dagger.hilt.android.AndroidEntryPoint;

import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;
import java.util.concurrent.TimeUnit;
import java.util.stream.Collectors;

@AndroidEntryPoint
public class AlarmFragment extends Fragment {

    private AlarmsAdapter adapter;
    private FragmentAlarmBinding binding;
    private Future<?> updateJob;
    private int interval = 0;
    private String wallpaper = "";
    private Map<Long, Long> millisToAlarm = new HashMap<>();
    private AlarmViewModel alarmViewModel;
    private final ExecutorService executorService = Executors.newSingleThreadExecutor();

    @SuppressLint("DiscouragedApi")
    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        alarmViewModel = new ViewModelProvider(this).get(AlarmViewModel.class);
        binding = FragmentAlarmBinding.inflate(inflater, container, false);
        alarmViewModel.getPreferencesWallpaperAndInterval(requireContext(), new PreferenceCallback() {
            @Override
            public void onResult(Pair<String, Integer> result) {
                wallpaper = result.first;
                interval = result.second;
            }
        });

        if (!Objects.equals(wallpaper, "")) {
            int resId = requireContext().getResources().getIdentifier(wallpaper, "drawable", requireContext().getPackageName());
            if (resId != 0) {
                binding.alarmLayout.setBackground(ContextCompat.getDrawable(requireContext(), resId));
            }
        }

        adapter = new AlarmsAdapter(interval, new AlarmActionListener() {
            @Override
            public void onAlarmEnabled(Alarm alarm, int index) {
                changeAlarmTime(alarm);
                binding.barTextView.setText(updateBar());

                alarmViewModel.updateEnabledAlarm(alarm, !alarm.getEnabled(), result -> {
                    if(result) {
                        adapter.notifyItemChanged(index);
                    }
                });
            }

            @Override
            public void onAlarmChange(Alarm alarm) {
                BottomSheetFragment fragment = new BottomSheetFragment(alarmViewModel,false, alarm, new BottomSheetListener() {
                    @Override
                    public void onAddAlarm(Alarm alarm) {
                    }

                    @Override
                    public void onChangeAlarm(Alarm alarmOld, Alarm alarmNew) {
                        if (alarmNew.getEnabled()) {
                            changeAlarmTime(alarmOld);
                            changeAlarmTime(alarmNew);
                            binding.barTextView.setText(updateBar());
                        }
                    }
                });
                fragment.show(getChildFragmentManager(), "ChangeTag");
            }

            @Override
            public void onAlarmLongClicked() {
                binding.floatingActionButtonAdd.setVisibility(View.GONE);
                binding.floatingActionButtonDelete.setVisibility(View.VISIBLE);
                requireActivity().getOnBackPressedDispatcher().addCallback(getViewLifecycleOwner(), new OnBackPressedCallback(true) {
                    @Override
                    public void handleOnBackPressed() {
                        if (!adapter.canLongClick) {
                            adapter.clearPositions();
                            adapter.notifyDataSetChanged();
                            binding.floatingActionButtonDelete.setVisibility(View.GONE);
                            binding.floatingActionButtonAdd.setVisibility(View.VISIBLE);
                            alarmViewModel.getAndNotify();
                        } else {
                            remove();
                            requireActivity().getOnBackPressedDispatcher().onBackPressed();
                        }
                    }
                });

                binding.floatingActionButtonDelete.setOnClickListener(v -> {
                    List<Alarm> alarmsToDelete = adapter.getDeleteList();
                    if (!alarmsToDelete.isEmpty()) {
                        alarmViewModel.deleteAlarms(alarmsToDelete, getContext());
                        for (Alarm a : alarmsToDelete) {
                            if (a.getEnabled()) changeAlarmTime(a);
                        }
                        binding.barTextView.setText(updateBar());
                        binding.floatingActionButtonDelete.setVisibility(View.GONE);
                        binding.floatingActionButtonAdd.setVisibility(View.VISIBLE);
                        adapter.clearPositions();
                    }
                });
            }
        });

        LinearLayoutManager layoutManager = new LinearLayoutManager(requireContext());
        binding.recyclerview.setLayoutManager(layoutManager);
        binding.recyclerview.setAdapter(adapter);
        binding.recyclerview.addItemDecoration(new VerticalSpaceItemDecoration(40));

        alarmViewModel.alarms.observe(getViewLifecycleOwner(), alarms -> adapter.setAlarms(alarms));

        ((AppCompatActivity) requireActivity()).setSupportActionBar(binding.toolbar);

        binding.floatingActionButtonAdd.setOnClickListener(v -> {
            BottomSheetFragment fragment = new BottomSheetFragment(alarmViewModel,true, new Alarm(0), new BottomSheetListener() {
                @Override
                public void onAddAlarm(Alarm alarm) {
                    long id = 0;
                    for (Alarm a : adapter.alarms) {
                        if (a.getTimeHours() == alarm.getTimeHours() && a.getTimeMinutes() == alarm.getTimeMinutes()) {
                            id = a.getId();
                            break;
                        }
                    }
                    Alarm alr = new Alarm(id);
                    alr.setTimeHours(alarm.getTimeHours());
                    alr.setTimeMinutes(alarm.getTimeMinutes());
                    alr.setName(alarm.getName());
                    alr.setEnabled(alarm.getEnabled());
                    changeAlarmTime(alr);
                    binding.barTextView.setText(updateBar());
                }

                @Override
                public void onChangeAlarm(Alarm alarmOld, Alarm alarmNew) {
                }
            });
            fragment.show(getChildFragmentManager(), "AddTag");
        });

        alarmViewModel.getInitCompleted().observe(getViewLifecycleOwner(), it -> {
            if (it) {
                updateJob = executorService.submit(() -> {
                    try {
                        millisToAlarm = fillAlarmsTime();
                        while (!Thread.currentThread().isInterrupted()) {
                            binding.barTextView.post(() -> binding.barTextView.setText(updateBar()));
                            TimeUnit.SECONDS.sleep(30);
                        }
                    } catch (InterruptedException e) {
                        Thread.currentThread().interrupt();
                    }
                });
            }
        });

        return binding.getRoot();
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        if (updateJob != null) {
            updateJob.cancel(true);
        }
    }

    private Map<Long, Long> fillAlarmsTime() {
        Map<Long, Long> map = new HashMap<>();
        Calendar calendar = Calendar.getInstance();
        Calendar calendar2 = Calendar.getInstance(ULocale.ROOT);

        for (Alarm alr : adapter.alarms) {
            if (alr.getEnabled()) {
                calendar.set(Calendar.HOUR_OF_DAY, alr.getTimeHours());
                calendar.set(Calendar.MINUTE, alr.getTimeMinutes());
                calendar.set(Calendar.SECOND, 0);
                long longTime = (calendar2.getTimeInMillis() > calendar.getTimeInMillis())
                        ? calendar.getTimeInMillis() + 86400000
                        : calendar.getTimeInMillis();
                map.put(alr.getId(), longTime);
            }
        }

        return map.entrySet().stream()
                .sorted(Map.Entry.comparingByValue())
                .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue, (e1, e2) -> e1, LinkedHashMap::new));
    }

    private void changeAlarmTime(Alarm alarm) {
        Log.d("testChangeTimeBefore", millisToAlarm.toString());
        if (millisToAlarm.containsKey(alarm.getId())) {
            Log.d("testRemove", String.valueOf(alarm.getId()));
            millisToAlarm.remove(alarm.getId());
        } else {
            Calendar calendar = Calendar.getInstance();
            Calendar calendar2 = Calendar.getInstance(ULocale.ROOT);
            calendar.set(Calendar.HOUR_OF_DAY, alarm.getTimeHours());
            calendar.set(Calendar.MINUTE, alarm.getTimeMinutes());
            calendar.set(Calendar.SECOND, 0);
            long longTime = (calendar2.getTimeInMillis() > calendar.getTimeInMillis())
                    ? calendar.getTimeInMillis() + 86400000
                    : calendar.getTimeInMillis();
            millisToAlarm.put(alarm.getId(), longTime);
        }
        Log.d("testChangeTimeMid", millisToAlarm.toString());
        millisToAlarm = millisToAlarm.entrySet().stream()
                .sorted(Map.Entry.comparingByValue())
                .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue, (e1, e2) -> e1, LinkedHashMap::new));
        Log.d("testChangeTimeAfter", millisToAlarm.toString());
    }

    private String updateBar() {
        StringBuilder txt = new StringBuilder();
        if (millisToAlarm.isEmpty()) {
            txt.append("Все сигналы\nвыключены");
        } else {
            Calendar calendar = Calendar.getInstance(ULocale.ROOT);
            long longTime = millisToAlarm.entrySet().iterator().next().getValue();
            int minutes = (int) ((longTime - calendar.getTimeInMillis()) / 60000);

            if (minutes == 0) {
                txt.append("Звонок менее чем через 1 мин.");
            } else if (minutes < 60) {
                txt.append("Звонок через\n").append(minutes).append(" мин.");
            } else {
                int hours = minutes / 60;
                txt.append("Звонок через\n").append(hours).append(" ч. ").append(minutes % 60).append(" мин.");
            }
        }
        Log.d("testUpdateBar", txt.toString());
        return txt.toString();
    }

    public void fillAndUpdateBar() {
        millisToAlarm = fillAlarmsTime();
        binding.barTextView.setText(updateBar());
    }

    @Override
    public void onPause() {
        super.onPause();
        if (updateJob != null) {
            updateJob.cancel(true); // Принудительно завершаем задачу
        }
    }
}

File: app/src/main/java/com/example/alarm/AlarmFragment.kt
package com.example.alarm

import android.annotation.SuppressLint
import android.content.Context
import android.content.SharedPreferences
import android.graphics.Rect
import android.icu.util.Calendar
import android.icu.util.ULocale
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.activity.OnBackPressedCallback
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.example.alarm.databinding.FragmentAlarmBinding
import com.example.alarm.model.Alarm
import com.example.alarm.model.AlarmService
import com.example.alarm.model.AlarmsListener
import com.example.alarm.model.MyAlarmManager
import com.example.alarm.model.Settings
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.async
import kotlinx.coroutines.delay
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

class AlarmFragment : Fragment() {

    private lateinit var adapter: AlarmsAdapter
    private lateinit var binding: FragmentAlarmBinding
    private var updateJob: Job? = null

    private var millisToAlarm = mutableMapOf<Long, Long>()
    private lateinit var alarmViewModel: AlarmViewModel
    @SuppressLint("DiscouragedApi")
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        binding = FragmentAlarmBinding.inflate(inflater, container, false)
        alarmViewModel = ViewModelProvider(requireActivity())[AlarmViewModel::class.java]
                val (wallpaper, interval) = alarmViewModel.getPreferencesWallpaperAndInterval(requireContext())
                if(wallpaper != "") {
                    val resId = resources.getIdentifier(wallpaper, "drawable", requireContext().packageName)
                    if(resId != 0)
                        binding.alarmLayout.background = ContextCompat.getDrawable(requireContext(), resId)
                }
                val settings = Settings(id = 0, interval = interval)
                adapter = AlarmsAdapter(settings, object : AlarmActionListener {
                    override fun onAlarmEnabled(alarm: Alarm, index: Int) {
                        lifecycleScope.launch {
                            var bool = 0
                            if (alarm.enabled == 0) {
                                bool = 1
                                changeAlarmTime(alarm, false)
                                binding.barTextView.text = updateBar()
                            } else {
                                changeAlarmTime(alarm, true)
                                binding.barTextView.text = updateBar()
                            }
                            val idx = async(Dispatchers.IO) {
                                alarmViewModel.updateEnabledAlarm(alarm, bool, requireContext(), index)
                            }
                            adapter.notifyItemChanged(idx.await()) //костыль, но по-другому нет идей как сделать
                        }
                    }
                    override fun onAlarmChange(alarm: Alarm) {
                        BottomSheetFragment(false, alarm, object : BottomSheetListener {
                            override fun onAddAlarm(alarm: Alarm) {
                                return
                            }

                            override fun onChangeAlarm(alarmOld: Alarm, alarmNew: Alarm) {
                                if (alarmNew.enabled == 1) {
                                    changeAlarmTime(alarmOld, true)
                                    changeAlarmTime(alarmNew, false)
                                    binding.barTextView.text = updateBar()
                                }
                            }
                        }).show(childFragmentManager, "ChangeTag")
                    }

                    override fun onAlarmLongClicked() {
                        binding.floatingActionButtonAdd.visibility = View.GONE
                        binding.floatingActionButtonDelete.visibility = View.VISIBLE
                        requireActivity().onBackPressedDispatcher.addCallback(
                            viewLifecycleOwner,
                            object : OnBackPressedCallback(true) {
                                @SuppressLint("NotifyDataSetChanged")
                                override fun handleOnBackPressed() {
                                    if (!adapter.canLongClick) {
                                        adapter.clearPositions()
                                        adapter.notifyDataSetChanged()
                                        binding.floatingActionButtonDelete.visibility = View.GONE
                                        binding.floatingActionButtonAdd.visibility = View.VISIBLE
                                        alarmViewModel.getAndNotify()
                                    } else {
                                        //Removing this callback
                                        remove()
                                        requireActivity().onBackPressedDispatcher.onBackPressed()
                                    }
                                }
                            })
                        binding.floatingActionButtonDelete.setOnClickListener {
                            val alarmsToDelete = adapter.getDeleteList()
                            if (alarmsToDelete.isNotEmpty()) {
                                    alarmViewModel.deleteAlarms(alarmsToDelete, context)
                                    for (a in alarmsToDelete) {
                                        if (a.enabled == 1) changeAlarmTime(a, true)
                                    }
                                    binding.barTextView.text = updateBar()
                                binding.floatingActionButtonDelete.visibility = View.GONE
                                binding.floatingActionButtonAdd.visibility = View.VISIBLE
                                adapter.clearPositions()

                            }
                        }
                    }
                })
                val layoutManager = LinearLayoutManager(requireContext())
                binding.recyclerview.layoutManager = layoutManager
                binding.recyclerview.adapter = adapter
                binding.recyclerview.addItemDecoration(VerticalSpaceItemDecoration(40))
                alarmViewModel.alarms.observe(viewLifecycleOwner) {
                    adapter.alarms = it
                }
                (activity as AppCompatActivity?)!!.setSupportActionBar(binding.toolbar) //adds a button
                binding.floatingActionButtonAdd.setOnClickListener {
                    BottomSheetFragment(true, Alarm(0), object : BottomSheetListener {
                        override fun onAddAlarm(alarm: Alarm) {
                                var id: Long = 0
                                for (a in adapter.alarms) {
                                    if (a.timeHours == alarm.timeHours && a.timeMinutes == alarm.timeMinutes) {
                                        id = a.id
                                        break
                                    }
                                }
                                val alr = Alarm(
                                    id = id,
                                    timeHours = alarm.timeHours,
                                    timeMinutes = alarm.timeMinutes,
                                    name = alarm.name,
                                    enabled = alarm.enabled
                                )
                                changeAlarmTime(alr, false)
                                binding.barTextView.text = updateBar()
                        }

                        override fun onChangeAlarm(alarmOld: Alarm, alarmNew: Alarm) {
                            return
                        }
                    }).show(childFragmentManager, "AddTag")
                }
        alarmViewModel.initCompleted.observe(viewLifecycleOwner) {
            Log.d("test","Works!")
            if (it) {
                updateJob = lifecycleScope.launch {
                    millisToAlarm = fillAlarmsTime()
                    while (isActive) {
                        binding.barTextView.text = updateBar()
                        delay(30000)
                    }
                }
            }
        }
        return binding.root
    }

    override fun onDestroyView() {
        super.onDestroyView()
        updateJob?.cancel()
    }

    private suspend fun fillAlarmsTime() : MutableMap<Long, Long> = withContext(Dispatchers.Default){
        val map = mutableMapOf<Long, Long>()
        val calendar = Calendar.getInstance()
        val calendar2 = Calendar.getInstance(ULocale.ROOT)
        for(alr in adapter.alarms) {
            if(alr.enabled == 1) {
                calendar.set(Calendar.HOUR_OF_DAY, alr.timeHours)
                calendar.set(Calendar.MINUTE, alr.timeMinutes)
                calendar.set(Calendar.SECOND, 0)
                val longTime: Long = if (calendar2.timeInMillis > calendar.timeInMillis) {
                    calendar.timeInMillis + 86400000
                } else calendar.timeInMillis
                map[alr.id] = longTime
            }
        }
        val sortedMap = map.toList().sortedBy { it.second }.toMap().toMutableMap()
        return@withContext sortedMap
    }

     private fun changeAlarmTime(alarm: Alarm, isDisable: Boolean) {
        if(isDisable) {
            millisToAlarm.remove(alarm.id)
        }
        else {
            val calendar = Calendar.getInstance()
            val calendar2 = Calendar.getInstance(ULocale.ROOT)
            calendar.set(Calendar.HOUR_OF_DAY, alarm.timeHours)
            calendar.set(Calendar.MINUTE, alarm.timeMinutes)
            calendar.set(Calendar.SECOND, 0)
            val longTime: Long = if (calendar2.timeInMillis > calendar.timeInMillis) {
                calendar.timeInMillis + 86400000
            } else calendar.timeInMillis
            millisToAlarm[alarm.id] = longTime
            millisToAlarm = millisToAlarm.toList().sortedBy { it.second }.toMap().toMutableMap()
        }
    }
    private fun updateBar(): String {
        Log.d("testBar", millisToAlarm.toString())
        var txt: String = ""
        if(millisToAlarm.isEmpty()) txt += "Все сигналы\nвыключены"
        else {
            val calendar = Calendar.getInstance(ULocale.ROOT)
            val longTime: Long = millisToAlarm.entries.first().value
            val minutes: Int = if (calendar.timeInMillis > longTime) {
                ((longTime - calendar.timeInMillis) / 60000).toInt()
            } else ((longTime - calendar.timeInMillis) / 60000).toInt()
            when(minutes) {
                0 -> txt += "Звонок менее чем через 1 мин."
                in 1..59 -> txt += "Звонок через\n$minutes мин."
                else -> {
                    val hours = minutes / 60
                    txt += "Звонок через\n$hours ч. ${minutes % 60} мин."
                }
            }
        }
        return txt
    }

    suspend fun fillAndUpdateBar() = withContext(Dispatchers.Default) {
        millisToAlarm = fillAlarmsTime()
        binding.barTextView.text = updateBar()
    }

    override fun onPause() {
        super.onPause()
        updateJob?.cancel()
    }
}

class VerticalSpaceItemDecoration(private val verticalSpaceHeight: Int) : RecyclerView.ItemDecoration() {
    override fun getItemOffsets(outRect: Rect, view: View, parent: RecyclerView, state: RecyclerView.State) {
        super.getItemOffsets(outRect, view, parent, state)
        outRect.bottom = verticalSpaceHeight
    }
}